Source/Core/GoogleEarthEnterpriseTerrainProvider.js

```diff
@@ -0,0 +1,572 @@
+        var parentInfo;
+        var q = quadKey;
+        var terrainVersion = -1;
+        var terrainState = info.terrainState;
+        if (terrainState !== TerrainState.NONE) {
+            switch (terrainState) {
+                case TerrainState.SELF: // We have terrain and have retrieved it before
+                    terrainVersion = info.terrainVersion;
+                    break;
+                case TerrainState.PARENT: // We have terrain in our parent
+                    q = q.substring(0, q.length - 1);
+                    parentInfo = metadata.getTileInformationFromQuadKey(q);
+                    terrainVersion = parentInfo.terrainVersion;
+                    break;
+                case TerrainState.UNKNOWN: // We haven't tried to retrieve terrain yet
+                    if (info.hasTerrain()) {
+                        terrainVersion = info.terrainVersion; // We should have terrain
+                    } else {
+                        q = q.substring(0, q.length - 1);
+                        parentInfo = metadata.getTileInformationFromQuadKey(q);
+                        if (defined(parentInfo) && parentInfo.hasTerrain()) {
+                            terrainVersion = parentInfo.terrainVersion; // Try checking in the parent
+                        }
+                    }
+                    break;
+            }
+        }
+
+        // We can't figure out where to get the terrain
+        if (terrainVersion < 0) {
+            return when.reject(new RuntimeError('Terrain tile doesn\'t exist'));
+        }
+
+        // Load that terrain
+        var terrainPromises = this._terrainPromises;
+        var url = buildTerrainUrl(this, q, terrainVersion);
+        var promise;
+        if (defined(terrainPromises[q])) { // Already being loaded possibly from another child, so return existing promise
+            promise = terrainPromises[q];
+        } else { // Create new request for terrain
+            var requestPromise;
+            throttleRequests = defaultValue(throttleRequests, true);
+            if (throttleRequests) {
+                requestPromise = throttleRequestByServer(url, loadArrayBuffer);
+                if (!defined(requestPromise)) {
+                    return undefined; // Throttled
+                }
+            } else {
+                requestPromise = loadArrayBuffer(url);
+            }
+
+            terrainPromises[q] = requestPromise;
+
+            promise = requestPromise
+                .always(function(terrain) {
+                    delete terrainPromises[q];
+                    return terrain;
+                });
+        }
+
+        return promise
+            .then(function(terrain) {
+                if (defined(terrain)) {
+                    var decodePromise = taskProcessor.scheduleTask({
+                        buffer : terrain,
+                        type : 'Terrain'
+                    });
+
+                    return decodePromise
```

Comment:
To avoid the extra indentation, whenever you return a promise in a chain, end the function and put a then one level up, instead of globbing then on this.  Essential, change this to

```javascript
return taskProcessor.scheduleTask({
                        buffer : terrain,
                        type : 'Terrain'
                    });
})
.then(function(terrainTiles) {
```